### PR TITLE
Migrate AST namespace to PHP 8 syntax

### PR DESCRIPTION
**Pull request by greg0ire**

### CODE REVIEW

File: lib/Doctrine/ORM/Query/AST/AggregateExpression.php [hunk 1]
```diff
@@ -8,21 +8,16 @@
 
 class AggregateExpression extends Node
 {
-    /**
-     * Some aggregate expressions support distinct, eg COUNT.
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query/AST/AggregateExpression.php:12]: We should keep this piece of documentation by moving it to the `@param` annotation.

File: lib/Doctrine/ORM/Query/AST/WhenClause.php [hunk 2]
```diff
@@ -13,11 +13,8 @@
  */
 class WhenClause extends Node
 {
-    /**
-     * @param ConditionalExpression $caseConditionExpression
-     * @param mixed                 $thenScalarExpression
-     */
-    public function __construct(public $caseConditionExpression = null, public $thenScalarExpression = null)
+    /** @param ConditionalExpression $caseConditionExpression */
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query/AST/WhenClause.php:16]: Apparently, we allow this parameter to be `null`.

```suggestion
    /** @param ConditionalExpression|null $caseConditionExpression */
```

Should be backported to 2.13.x or 2.14.x probably.

**Review by derrabus — DISMISSED** [review]

File: lib/Doctrine/ORM/Query/AST/WhenClause.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query/AST/WhenClause.php:16]: it's incomplete. 2.13.x is for wrong stuff, 2.14.x for imprecise stuff. I think incomplete here is wrong: it doesn't have to always be a `ConditionalExpression`

Comment (derrabus) [reply lib/Doctrine/ORM/Query/AST/WhenClause.php:16]: Looking at #10224: The `$caseConditionExpression` parameter was not optional on 2.x. Why is it here? Let's remove the default and document it as not nullable instead?

Comment (greg0ire) [reply lib/Doctrine/ORM/Query/AST/WhenClause.php:16]: :thinking: indeed. I think that must be a choice rector makes between the doc of the property, and the doc of the parameter. I'll amend #10224 after confirming it's safe to do so

Comment (greg0ire) [reply lib/Doctrine/ORM/Query/AST/WhenClause.php:16]: Using `object` as a type hint passes the test suite (and `ConditionalExpression` does not, of course lol)

**Comment by greg0ire** [comment]
Blocked by https://github.com/doctrine/orm/pull/10224

File: lib/Doctrine/ORM/Query/AST/BetweenExpression.php [hunk 3]
```diff
@@ -8,28 +8,18 @@
...
-
-    /** @var ArithmeticExpression */
-    public $leftBetweenExpression;
-
-    /** @var ArithmeticExpression */
-    public $rightBetweenExpression;
-
-    /** @var bool */
-    public $not;
+    public bool $not;
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query/AST/BetweenExpression.php]:
```suggestion
    public bool $not = false;
```

We should not allow this property to be uninitialized. In fact, I'd suggest to add it as promoted property to the constructor. `Parser::BetweenExpression()` currently assigns it right after construction anyway.

See #10267.

File: lib/Doctrine/ORM/Query/Parser.php [hunk 4]
```diff
@@ -1007,7 +1007,7 @@ public function AliasIdentificationVariable()
     /**
      * AbstractSchemaName ::= fully_qualified_name | identifier
      *
-     * @return string
+     * @return class-string
```
Comment (derrabus) [inline lib/Doctrine/ORM/Query/Parser.php]: The implementation looks like we don't always get a FQCN here. I'd revert this change.

**Review by derrabus — CHANGES_REQUESTED** [review]
Rebase needed.

**Comment by greg0ire** [comment]
Converting to draft until your changes on this namespace are merged up :slightly_smiling_face:

**Review by derrabus — APPROVED** [review]

**Pull request merged**